server/src/main/java/org/elasticsearch/search/query/SingleThreadCollectorManager.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.query;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/** {@link CollectorManager} implementation that assumes it is running non-concurrently. Therefore
+ * {@link #newCollector()} will be called only once. */
+public abstract class SingleThreadCollectorManager implements CollectorManager<Collector, Void> {
+
+    private Collector collector;
```

Comment:
The point of this class is to make sure newCollector is only called once, and the reduce method is called with exactly the object that was returned by this method. It has save a couple of mistakes where the method was called more than once or reduce was called with incorrect objects.

Reply:
you can leave the assertion in the newCollector method, but you can pre-compute the collector and provide it at construction time instead of requiring to extend the class and override the getCollector method.

Reply:
I change the class to be concrete and final, let me know what you think.